crates/dslab-core/src/async_core/shared_state.rs

```diff
@@ -0,0 +1,352 @@
+    ///
+    /// root_ctx.emit_now(Start {}, client_id);
+    /// root_ctx.emit(SomeEvent { request_id: 1 }, client_id, 50.);
+    /// root_ctx.emit(SomeEvent { request_id: 2 }, client_id, 60.);
+    ///
+    /// sim.step_until_no_events();
+    ///
+    /// assert_eq!(*client.borrow().actions_finished.borrow(), 2);
+    /// assert_eq!(sim.time(), 110.); // because of timers in listen_first
+    /// ```
+    pub async fn with_timeout(self, timeout: f64) -> AwaitResult<T> {
+        assert!(timeout >= 0., "timeout must be a positive value");
+
+        let component_id = self.state.borrow().component_id;
+        self.sim_state
+            .borrow_mut()
+            .add_timer_on_state(component_id, timeout, self.state.clone());
+        EventWithTimeoutFuture { state: self.state }.await
+    }
+}
+
+/// Future that represents timer from simulation.
+pub struct TimerFuture {
+    /// State that should be completed after timer fired.
+    pub(crate) state: Rc<RefCell<AwaitEventSharedState<EmptyData>>>,
+}
+
+impl Future for TimerFuture {
+    type Output = ();
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        let mut state = self.state.as_ref().borrow_mut();
+
+        if !state.completed {
+            state.waker = Some(cx.waker().clone());
+            return Poll::Pending;
+        }
+
+        Poll::Ready(())
+    }
+}
+
+#[derive(Hash, PartialEq, Eq, Debug, Clone, Copy)]
+pub(crate) struct AwaitKey {
+    pub to: Id,
+    pub msg_type: TypeId,
+    event_key: EventKey,
+}
+
+impl AwaitKey {
+    pub fn new<T: EventData>(to: Id) -> Self {
+        Self {
+            to,
+            msg_type: TypeId::of::<T>(),
+            event_key: 0,
+        }
+    }
+
+    pub fn new_by_ref(to: Id, data: &dyn EventData) -> Self {
+        Self {
+            to,
+            msg_type: data.type_id(),
+            event_key: 0,
+        }
+    }
+
+    pub fn new_with_event_key<T: EventData>(to: Id, event_key: EventKey) -> Self {
+        Self {
+            to,
+            msg_type: TypeId::of::<T>(),
+            event_key,
+        }
+    }
+
+    pub fn new_with_event_key_by_ref(to: Id, data: &dyn EventData, event_key: EventKey) -> Self {
+        Self {
+            to,
+            msg_type: data.type_id(),
+            event_key,
+        }
+    }
+}
+
+struct EventWithTimeoutFuture<T: EventData> {
+    pub state: Rc<RefCell<AwaitEventSharedState<T>>>,
+}
+
+impl<T: EventData> Future for EventWithTimeoutFuture<T> {
+    type Output = AwaitResult<T>;
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        let mut state = self.state.as_ref().borrow_mut();
+
+        if !state.completed {
+            state.waker = Some(cx.waker().clone());
+            return Poll::Pending;
+        }
+
+        let mut filler = AwaitResult::default();
+        std::mem::swap(&mut filler, &mut state.shared_content);
+
+        Poll::Ready(filler)
```

Comment:
Кажется это эквивалентно

```suggestion
        Poll::Ready(std::mem::take(&mut state.shared_content))
```

Reply:
заменил